### DISCUSSION TITLE
How can i call PopupWindow‘s show() when hover within TouchArea

### DISCUSSION THREAD

**Discussion by wuanzhuan** [slint-ui/slint#4395] [Q&A]
I need show a PopupWindow when hover, and close the PopupWindow when leave hover.
    but i can't call the show() when the TouchArea's has-hover is changing.  Anyway, how can i do?

    popup := PopupWindow {
        Rectangle {
            Text {
                text: "example";
            }
        }
    }

    i-touch-area := TouchArea {
        // like this
        has-hover(is_hover) => {
            if (is_hover) {
                popup.show();
            } else {
                popup.close();
        }
    }

**Comment by ogoffart** [comment]
In the master branch (which will be released as part of Slint 1.4), you can use the PointerEventKind.move in the pointer-event callback.

[Example](https://slint.dev/snapshots/master/editor/?snippet=import+%7B+AboutSlint%2C+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0Aexport+component+Demo+%7B%0A++++width%3A+300px%3B%0A++++height%3A+300px%3B%0A++++Rectangle+%7B%0A++++++++background%3A+green%3B%0A++++++++width%3A+20%25%3B+height%3A+20%25%3B%0A++++++++TouchArea+%7B%0A++++++++++++pointer-event%28ev%29+%3D%3E+%7B%0A++++++++++++++++if+%28ev.kind+%3D%3D+PointerEventKind.move%29+%7B%0A++++++++++++++++++++popup.show%28%29%3B%0A++++++++++++++++%7D%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A%0A++++popup+%3A%3D+PopupWindow+%7B%0A++++++++width%3A+200px%3B%0A++++++++height%3A+200px%3B%0A++++++++Rectangle+%7B%0A++++++++++++background%3A+yellow%3B%0A++++++++%7D%0A++++%7D%0A%7D%0A)

```
        TouchArea {
            pointer-event(ev) => {
                if (ev.kind == PointerEventKind.move) {
                    popup.show();
                }
            }
        }
```

**Reply by ogoffart** [reply]
Note that PopupWindow might not be the best for this kind of things because it also prevent events to be forwareded.
If what you want to do is some kind of tooltip, there are some discussion on https://github.com/slint-ui/slint/discussions/1617#discussioncomment-6670716

**Reply by bjorn** [reply]
Yeah, the `move` event is not enough, since we also need the `leave` event to be able to close the popup (and of course, the popup should in this case ignore or pass-through mouse events).

A `PopupWindow` is as far as I am aware the only way to position a widget on top of the rest of the hierarchy. In contrast, it is very tricky to position any other kind of widget both relative to the hovered item as well as on top of the widget hierarchy, especially if it needs to be positioned relative to a list item.

**Reply by ogoffart** [reply]
the `move` event is only called if you press
In retrospect, it shouldn't have been like that but we didn't want to change the behavior.

> A PopupWindow is as far as I am aware the only way to position a widget on top of the rest of the hierarchy. In contrast, it is very tricky to position any other kind of widget both relative to the hovered item as well as on top of the widget hierarchy, 

True that's a good point.

**Reply by bjorn** [reply]
> the `move` event is only called if you press

Are you sure? Your own example above demonstrates one does not need to press when hovering the green square, to spawn the yellow one.

**Reply by wuanzhuan** [reply]
I need the PopupWindow that can be above and outside the TouchArea. and can trigger when hover. the #1617 is not same as Popup Window. 
the #1617:
![image](https://github.com/slint-ui/slint/assets/11628049/e220fc33-db3d-440d-850e-c65bf02ae69f)
i need:
![image](https://github.com/slint-ui/slint/assets/11628049/b2e68e75-63da-48f6-a644-25287d14c5eb)

**Reply by ogoffart** [reply]
> Are you sure? 

My mistake, i confused the `move` event with `moved` callback.

> I need the PopupWindow that can be above and outside the TouchArea

Right, you will need a PopupWindow for that.

**Reply by wuanzhuan** [reply]
@ogoffart how can i custom the buildin window liking popupwindow. It can be above and outside the TouchArea.
now the window can't show. i set the properties of z:1000 or always-on-top:true.
The one way seems to use a global that has a has-hover. and create my popupwindows by root element when the global's has-hover is true. but this is not elegant
